app/controllers/pensions/annualAllowance/PensionSchemeTaxReferenceController.scala

```diff
@@ -0,0 +1,111 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.pensions.annualAllowance
+
+import config.{AppConfig, ErrorHandler}
+import controllers.predicates.AuthorisedAction
+import forms.PensionSchemeTaxReferenceForm
+import models.mongo.PensionsCYAModel
+import play.api.data.Form
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import services.PensionSessionService
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendController
+import utils.Clock
+import views.html.pensions.annualAllowance.PensionSchemeTaxReferenceView
+import controllers.pensions.routes.PensionsSummaryController
+import models.User
+
+import javax.inject.Inject
+import scala.concurrent.Future
+
+class PensionSchemeTaxReferenceController @Inject()(implicit val cc: MessagesControllerComponents,
+                                                    authAction: AuthorisedAction,
+                                                    pensionSchemeTaxReferenceView: PensionSchemeTaxReferenceView,
+                                                    appConfig: AppConfig,
+                                                    pensionSessionService: PensionSessionService,
+                                                    errorHandler: ErrorHandler,
+                                                    clock: Clock) extends FrontendController(cc) with I18nSupport {
+
+
+  private def prefillValue(pstrListOpt: Option[Seq[String]], pensionSchemeTaxReference: Option[Int], user: User): Form[String] = {
+    val emptyForm: Form[String] = PensionSchemeTaxReferenceForm.pensionSchemeTaxReferenceForm(user.isAgent)
+    (pstrListOpt, pensionSchemeTaxReference) match {
+      case (Some(pstrList), Some(pstrIndex)) => if (pstrList.size < pstrIndex) emptyForm else emptyForm.fill(pstrList(pstrIndex))
+      case (None, Some(_)) =>
+        //Person has request a certain pstr but it doesn't exist on the list, this shouldn't happen unless it's through an API request
+        //in this case the prefill value will be None.
+        emptyForm
+      case (Some(_), None) => emptyForm
+      case (None, None) => emptyForm
+    }
+  }
+
+  def show(taxYear: Int, pensionSchemeTaxReference: Option[Int]): Action[AnyContent] = authAction.async { implicit request =>
+
+    pensionSessionService.getPensionsSessionDataResult(taxYear, request.user) {
+      case Some(data) =>
+        val form = prefillValue(
+          data.pensions.pensionsAnnualAllowances.pensionSchemeTaxReference,
+          pensionSchemeTaxReference,
+          request.user
+        )
+        Future.successful(Ok(pensionSchemeTaxReferenceView(form, taxYear, pensionSchemeTaxReference)))
+      case _ =>
+        Future.successful(Redirect(PensionsSummaryController.show(taxYear)))
+    }
+  }
+
+  def submit(taxYear: Int, pensionSchemeTaxReference: Option[Int]): Action[AnyContent] = authAction.async {
```

Comment:
Could you rename the parameter as it's not actually a pensionSchemeTaxReference but an index
Either pensionSchemeTaxReferenceIndex of similar (pstrIndex or just index?)
The array is called pensionSchemeTaxReference too which makes the code confusing both being named the same